pkg/util/subnet/endpointsadder.go

```diff
@@ -0,0 +1,84 @@
+package subnet
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+import (
+	"strings"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-08-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+)
+
+//  EndpointsAdder is the interface that wraps the basic logic to add endpoints to subnets
+//
+// AddEndpointsToSubnets adds the endpoints (that either are missing in subnets
+// or aren't in succeded state in subnets) to the subnets and returns those subnets.
+// Implementations of this method should not talk to infrastructure layer (e.g, Azure).
+// In contrast, the result of this function should be passed to a subnet manager to update the subnets
+// in Azure.
+type EndpointsAdder interface {
```

Comment:
IMHO even interfaces are used most of the time to avoid talking directly to external dependencies (DB, HTTP, CLI, etc.), they are also used as a design mechanism to decouple components (in this case "enableServiceEndpoints") from details ("AddEndpointsToSubnets"). 

If you don't have it as an interface, how can you ensure when testing pkg/cluster/enableendpoints.go "enableServiceEndpoints" function that "AddEndpointsToSubnets(...)" returns nil (or a specific value), and depending on that value "enableServiceEndpoints" behaves in a specific way? In other words, how can you test the behavior of "enableServiceEndpoints" without depending on the behavior of "EndpointsAdderManager" (implementation of the interface EndpointsAdder)?

Without the interface,  tests of "enableServiceEndpoints" depend on the behavior of "endpointsAdder" as it calls a specific method of a struct (and you are out of control of that).

In order to test "enableServiceEndpoints" independently from "EndpintsAdder" we need to create an interface to expose the behavior of "EndpointsAdder" but not the details. Without this interface, for example, how can we force AddEndpointsToSubnets to return nil?

Reply:
here your design is the problem. 
```go
func (m *manager) enableServiceEndpoints(ctx context.Context) error {
	subnetIds, err := m.getSubnetIds()
	if err != nil {
		return err
	}

	subnets, err := m.subnet.GetAll(ctx, subnetIds)
	if err != nil {
		return err
	}

	subnetsToBeUpdated := m.endpointsAdder.AddEndpointsToSubnets(api.SubnetsEndpoints, subnets)

	return m.updateSubnets(subnetsToBeUpdated, ctx)
}
```

you should not have to call another function from the subnet package. (it's hidden because you have the interface, but still) `m.endpointsAdder.AddEndpointsToSubnets(api.SubnetsEndpoints, subnets)`.  

The step where you add the subnets should be inside `m.updateSubnets()` . whether it is inside createOrUpdate or another function, it's up to you, but this step should be in the `subnet` package, not the `cluster` package.

Reply:
I didn't mean that as a mean comment. It is not a very big problem, true, but it still impacts the ability to test easily and to understand the code, so my statement was true nonetheless. 

This line needs to go in subnet `subnetsToBeUpdated:=m.endpointsAdder.AddEndpointsToSubnets(api.SubnetsEndpoints, subnets)`. and please remove the interface, it is not serving any purpose here.

Reply:
Done.